src/core/qgstracer.cpp

```diff
@@ -0,0 +1,676 @@
+{
+  // find edge where the point is
+  int lineVertexAfter;
+  int eIdx = point2edge( g, pt, lineVertexAfter );
+
+  //qDebug("e: %d", eIdx);
+
+  if ( eIdx == -1 )
+    return -1;
+
+  const QgsTracerGraph::E& e = g.e[eIdx];
+  QgsTracerGraph::V& v1 = g.v[e.v1];
+  QgsTracerGraph::V& v2 = g.v[e.v2];
+
+  QgsPolyline out1, out2;
+  split_linestring( e.coords, pt, lineVertexAfter, out1, out2 );
+
+  int vIdx = g.v.count();
+  int e1Idx = g.e.count();
+  int e2Idx = e1Idx + 1;
+
+  // prepare new vertex and edges
+
+  QgsTracerGraph::V v;
+  v.pt = pt;
+  v.edges << e1Idx << e2Idx;
+
+  QgsTracerGraph::E e1;
+  e1.v1 = e.v1;
+  e1.v2 = vIdx;
+  e1.coords = out1;
+
+  QgsTracerGraph::E e2;
+  e2.v1 = vIdx;
+  e2.v2 = e.v2;
+  e2.coords = out2;
+
+  // update edge connectivity of existing vertices
+  v1.edges.replace( v1.edges.indexOf( eIdx ), e1Idx );
+  v2.edges.replace( v2.edges.indexOf( eIdx ), e2Idx );
+  g.inactiveEdges << eIdx;
+
+  // add new vertex and edges to the graph
+  g.v.append( v );
+  g.e.append( e1 );
+  g.e.append( e2 );
+  g.joinedVertices++;
+
+  return vIdx;
+}
+
+
+int point_in_graph( QgsTracerGraph& g, const QgsPoint& pt )
+{
+  // try to use existing vertex in the graph
+  int v = point2vertex( g, pt );
+  if ( v != -1 )
+    return v;
+
+  // try to add the vertex to an edge (may fail if point is not on edge)
+  return join_vertex_to_graph( g, pt );
+}
+
+
+void reset_graph( QgsTracerGraph& g )
+{
+  // remove extra vertices and edges
+  g.v.resize( g.v.count() - g.joinedVertices );
+  g.e.resize( g.e.count() - g.joinedVertices * 2 );
+  g.joinedVertices = 0;
+
+  // fix vertices of deactivated edges
+  foreach ( int eIdx, g.inactiveEdges )
+  {
+    if ( eIdx >= g.e.count() )
+      continue;
+    const QgsTracerGraph::E& e = g.e[eIdx];
+    QgsTracerGraph::V& v1 = g.v[e.v1];
+    for ( int i = 0; i < v1.edges.count(); ++i )
+    {
+      if ( v1.edges[i] >= g.e.count() )
+        v1.edges.remove( i-- );
+    }
+    v1.edges << eIdx;
+
+    QgsTracerGraph::V& v2 = g.v[e.v2];
+    for ( int i = 0; i < v2.edges.count(); ++i )
+    {
+      if ( v2.edges[i] >= g.e.count() )
+        v2.edges.remove( i-- );
+    }
+    v2.edges << eIdx;
+  }
+
+  g.inactiveEdges.clear();
+}
+
+
+void extract_linework( QgsGeometry* g, QgsMultiPolyline& mpl )
+{
+  switch ( QgsWKBTypes::flatType( g->geometry()->wkbType() ) )
+  {
+    case QgsWKBTypes::LineString:
+      mpl << g->asPolyline();
+      break;
+
+    case QgsWKBTypes::Polygon:
+      foreach ( const QgsPolyline& ring, g->asPolygon() )
+        mpl << ring;
+      break;
+
+    case QgsWKBTypes::MultiLineString:
+      foreach ( const QgsPolyline& linestring, g->asMultiPolyline() )
+        mpl << linestring;
+      break;
+
+    case QgsWKBTypes::MultiPolygon:
+      foreach ( const QgsPolygon& polygon, g->asMultiPolygon() )
+        foreach ( const QgsPolyline& ring, polygon )
+          mpl << ring;
+      break;
+
+    default:
```

Comment:
It's trivial to handle this by using inheritance information from RTI and call the segmentize() function, so I'm not sure implementing this on iterator side is necessary.

https://github.com/qgis/QGIS/blob/master/src/core/qgsexpression.cpp#L1455